pkg/http/listener.go

```diff
@@ -0,0 +1,326 @@
+	updateBytesReadFunc    func(int)                           // function to be called to update bytes read in bufConn.
+	updateBytesWrittenFunc func(int)                           // function to be called to update bytes written in bufConn.
+	errorLogFunc           func(error, string, ...interface{}) // function to be called on errors.
+}
+
+// start - starts separate goroutine for each TCP listener.  A valid insecure/TLS HTTP new connection is passed to httpListener.acceptCh.
+func (listener *httpListener) start() {
+	listener.acceptCh = make(chan acceptResult)
+
+	// Closure to send acceptResult to acceptCh.
+	// It returns true if the result is sent else false if returns when doneCh is closed.
+	send := func(result acceptResult, doneCh <-chan struct{}) bool {
+		select {
+		case listener.acceptCh <- result:
+			// Successfully written to acceptCh
+			return true
+		case <-doneCh:
+			// As stop signal is received, close accepted connection.
+			if result.conn != nil {
+				result.conn.Close()
+			}
+			return false
+		}
+	}
+
+	// Closure to accept single connection.
+	acceptTCP := func(tcpConn *net.TCPConn, doneCh <-chan struct{}) {
+		// Tune accepted TCP connection.
+		tcpConn.SetKeepAlive(true)
+		tcpConn.SetKeepAlivePeriod(listener.tcpKeepAliveTimeout)
+
+		bufconn := newBufConn(tcpConn, listener.readTimeout, listener.writeTimeout,
+			listener.updateBytesReadFunc, listener.updateBytesWrittenFunc)
+
+		// Peek bytes of maximum length of all HTTP methods.
+		data, err := bufconn.Peek(methodMaxLen)
+		if err != nil {
+			if listener.errorLogFunc != nil {
+				listener.errorLogFunc(err,
+					"Error in reading from new connection %s at server %s",
+					bufconn.RemoteAddr(), bufconn.LocalAddr())
+			}
+			bufconn.Close()
+			return
+		}
+
+		// Return bufconn if read data is a valid HTTP method.
+		tokens := strings.SplitN(string(data), " ", 2)
+		if isHTTPMethod(tokens[0]) {
+			if listener.tlsConfig == nil {
+				send(acceptResult{bufconn, nil}, doneCh)
+			} else {
+				// As TLS is configured and we got plain text HTTP request,
+				// return 403 (forbidden) error.
+				bufconn.Write(sslRequiredErrMsg)
+				bufconn.Close()
+			}
+			return
+		}
+
+		if listener.tlsConfig != nil {
+			// As the listener is configured with TLS, try to do TLS handshake, drop the connection if it fails.
+			tlsConn := tls.Server(bufconn, listener.tlsConfig)
+			if err := tlsConn.Handshake(); err != nil {
+				if listener.errorLogFunc != nil {
+					listener.errorLogFunc(err,
+						"TLS handshake failed with new connection %s at server %s",
+						bufconn.RemoteAddr(), bufconn.LocalAddr())
+				}
+				bufconn.Close()
+				return
+			}
+
+			// Check whether the connection contains HTTP request or not.
+			bufconn = newBufConn(tlsConn, listener.readTimeout, listener.writeTimeout,
+				listener.updateBytesReadFunc, listener.updateBytesWrittenFunc)
+
+			// Peek bytes of maximum length of all HTTP methods.
+			data, err := bufconn.Peek(methodMaxLen)
```

Comment:
Why is there a double Peek() ? there is a Peek() before.. Previous implementation has only one.